src/win32/waiter.cpp

```diff
@@ -92,41 +92,50 @@ void Waiter::bumpds()
 // network layer)
 int WinWaiter::wait()
 {
-    int r = 0;
-
     // only allow interaction of asynccallback() with the main process while
     // waiting (because WinHTTP is threaded)
     if (pcsHTTP)
     {
         LeaveCriticalSection(pcsHTTP);
     }
 
+    int r = 0;
     addhandle(externalEvent, NEEDEXEC);
-    DWORD dwWaitResult = WaitForMultipleObjectsEx((DWORD)index, &handles.front(), FALSE, maxds * 100, TRUE);
 
-    if (pcsHTTP)
+    if (index <= MAXIMUM_WAIT_OBJECTS)
     {
-        EnterCriticalSection(pcsHTTP);
-    }
+        DWORD dwWaitResult = WaitForMultipleObjectsEx((DWORD)index, &handles.front(), FALSE, maxds * 100, TRUE);
+        assert(dwWaitResult != WAIT_FAILED);
 
 #ifdef MEGA_MEASURE_CODE
-    if (dwWaitResult == WAIT_TIMEOUT && maxds > 0) ++performanceStats.waitTimedoutNonzero;
-    else if (dwWaitResult == WAIT_TIMEOUT && maxds == 0) ++performanceStats.waitTimedoutZero;
-    else if (dwWaitResult == WAIT_IO_COMPLETION) ++performanceStats.waitIOCompleted;
-    else if (dwWaitResult >= WAIT_OBJECT_0) ++performanceStats.waitSignalled;
+        if (dwWaitResult == WAIT_TIMEOUT && maxds > 0) ++performanceStats.waitTimedoutNonzero;
+        else if (dwWaitResult == WAIT_TIMEOUT && maxds == 0) ++performanceStats.waitTimedoutZero;
+        else if (dwWaitResult == WAIT_IO_COMPLETION) ++performanceStats.waitIOCompleted;
+        else if (dwWaitResult >= WAIT_OBJECT_0) ++performanceStats.waitSignalled;
 #endif
 
-    if ((dwWaitResult == WAIT_TIMEOUT) || (dwWaitResult == WAIT_IO_COMPLETION) || maxds == 0)
-    {
-        r = NEEDEXEC;
+        if ((dwWaitResult == WAIT_TIMEOUT) || (dwWaitResult == WAIT_IO_COMPLETION) || maxds == 0 || (dwWaitResult == WAIT_FAILED))
+        {
+            r |= NEEDEXEC;
+        }
+        if ((dwWaitResult >= WAIT_OBJECT_0) && (dwWaitResult < WAIT_OBJECT_0 + flags.size()))
+        {
+            r |= flags[dwWaitResult - WAIT_OBJECT_0];
+        }
     }
-    if ((dwWaitResult >= WAIT_OBJECT_0) && (dwWaitResult < WAIT_OBJECT_0 + flags.size()))
+    else
     {
-        r |= flags[dwWaitResult - WAIT_OBJECT_0];
+        assert(false); // alert developers that we're hitting the limit
+        r |= NEEDEXEC;
```

Comment:
If we need to wait, but we immediately tell the Waiter that we need execution.. :thinking: it would be risky, IMO. We will do a loop, no event may actually need execution, so we loop for nothing and next call to `wait()` will again `NEEDEXEC`. Additionally, in the case that one event has completed, the line above will not avoid the missing notification when the new event to wait for completes (so we end up in the same situation that the current observed issue). 

Please, note the above is how I understand it. Of course, I might be wrong, don't see the full picture in here... so please, let me know if I'm misunderstanding how it works. Thanks in advance.

On the other hand, now that we are not adding so many "wait objects" thanks to reusing the same one to wait for multiple sockets (cUrl HANDLEs), it should not be a problem anymore (or at least hard to hit the limit). So I kind of agree it's a pragmatic workaround, specially with the `assert(false)` in there :smiley:

Reply:
Yes, the idea here is that this code path should not be hit, but if we expand handle usage such that we start hitting the limit again, developers will notice first in testing due to the assert(), and at least we won't suffer from mysterious pauses.   If it does actually happen, then while we have > 64 handles, we will be in a busy loop, consuming one core, but again without the mysterious pauses, and the busy core would be something less mysterious for devs to investigate.   Once the number of handles falls below 64 again, the busy loop goes back to normal.